zerver/actions/streams.py

```diff
@@ -133,6 +133,67 @@ def do_deactivate_stream(
     )
 
 
+def bulk_delete_cache_keys(message_ids_to_clear: List[int]) -> None:
+    while len(message_ids_to_clear) > 0:
+        batch = message_ids_to_clear[0:5000]
+
+        keys_to_delete = [to_dict_cache_key_id(message_id) for message_id in batch]
+        cache_delete_many(keys_to_delete)
+
+        message_ids_to_clear = message_ids_to_clear[5000:]
+
+
+def merge_streams(realm: Realm, stream_to_keep: Stream, stream_to_destroy: Stream) -> List[int]:
+    recipient_to_destroy = stream_to_destroy.recipient
+    recipient_to_keep = stream_to_keep.recipient
+    if recipient_to_destroy.id == recipient_to_keep.id:
+        return [0, 0, 0]
+
+    # gather info before taking action
+    existing_subs = Subscription.objects.filter(recipient=recipient_to_keep)
+    users_already_subscribed = {sub.user_profile_id: sub.active for sub in existing_subs}
+    message_ids_to_clear = list(
+        Message.objects.filter(recipient=recipient_to_destroy).values_list("id", flat=True)
+    )
+    subs_to_deactivate = Subscription.objects.filter(recipient=recipient_to_destroy, active=True)
+    users_to_activate = [
+        sub.user_profile
+        for sub in subs_to_deactivate
+        if not users_already_subscribed.get(sub.user_profile_id, False)
+    ]
+
+    # follow a safe order of operations, see
+    # https://github.com/zulip/zulip/pull/22262#pullrequestreview-1017884964
```

Comment:
If we have to have a comment here arguing for why this order is right (and I'm not convinced we do), then we should inline that argument -- we shouldn't make people fire up their browser to understand why.

Reply:
DONE - yeah, upon rereading this, I agree it doesn't need an explanation.